### PR TITLE
Fix USB OTG forgetting uris when paused

### LINKED ISSUE #1175: Fix OTG Uri forgetting
USB OTG Uri is forgotten when MainActivity pauses.

### PR DESCRIPTION
**Pull request by EmmanuelMess**
Fixes #1175.

Test needs some work.

### CODE REVIEW

**Comment by EmmanuelMess** [comment]
Need help with test, checkout WIP commit.

File: app/src/main/java/com/amaze/filemanager/filesystem/usb/SingletonUsbOtg.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package com.amaze.filemanager.filesystem.usb;
+
+import android.bluetooth.BluetoothClass;
+import android.net.Uri;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+public class SingletonUsbOtg {
```
Comment (TranceLove) [inline app/src/main/java/com/amaze/filemanager/filesystem/usb/SingletonUsbOtg.java]: Prefer name as `UsbOtgSingleton`.

Comment (EmmanuelMess) [reply app/src/main/java/com/amaze/filemanager/filesystem/usb/SingletonUsbOtg.java]: That was discussed previously (can't find comment though, so [here is @VishalNehra telling me to correct this name]( https://github.com/TeamAmaze/AmazeFileManager/pull/1172#discussion_r181566394)).

**Comment by EmmanuelMess** [comment]
@TranceLove
Test stuck after:
```
 MainActivity activity = Robolectric.buildActivity(MainActivity.class)
                .create().start().visible().resume().get();
```
Any ideas?

File: app/src/main/java/com/amaze/filemanager/filesystem/usb/SingletonUsbOtg.java (hunk 1 above)
Comment (TranceLove) [reply app/src/main/java/com/amaze/filemanager/filesystem/usb/SingletonUsbOtg.java]: That's OK then.

**Comment by TranceLove** [comment]
> @TranceLove
Test stuck after:

```
MainActivity activity = Robolectric.buildActivity(MainActivity.class)
.create().start().visible().resume().get();
```

If all you need is get it running (but still fails), here's a better way to setup the Activity class. [Reference](http://robolectric.org/activity-lifecycle/)

```
MainActivity activity = Robolectric.buildActivity(MainActivity.class).create().get();
```

**Comment by EmmanuelMess** [comment]
> If all you need is get it running (but still fails), here's a better way to setup the Activity class. Reference

I need the onResume() to get called.

**Comment by TranceLove** [comment]
> I need the onResume() to get called.

```
ActivityController<MainActivity> controller = Robolectric.buildActivity(MainActivity.class).create();
MainActivity activity = controller.get();
controller.resume();
```
Regarding the test hang problem, when debug the test on API 26 I saw it was requesting storage permissions (`PermissionsActivity.requestStoragePermission()`) ... but I'm not sure if it's where you stuck at.

**Comment by EmmanuelMess** [comment]
This:
```
ActivityController<MainActivity> controller = Robolectric.buildActivity(MainActivity.class).create();
MainActivity activity = controller.get();
controller.resume();
```
Runs up to the line ```controller.resume();```, then gets stuck.

**Comment by TranceLove** [comment]
Yes, same here, but I'm stuck when test is running on API 26 - see above. For others,

API 19 failed with `java.lang.StackOverflowError`
API 21, 22 failed with `UsbDevice` constructor problem
API 23 - 25 failed with assertion failed

Now I'm keeping your code verbatim, but I explicitly asked for storage state as mounted before starting. Maybe can help you a bit?

```
ShadowEnvironment.setExternalStorageState(Environment.MEDIA_MOUNTED);
MainActivity activity = Robolectric.buildActivity(MainActivity.class)
    .create().start().visible().resume().get();
```

**Comment by EmmanuelMess** [comment]
I have no idea how to tell the system that a (real) directory should be used as a (simulated) USB OTG storage.

**Comment by EmmanuelMess** [comment]
Okay, I removed the test, no idea how to make it work, I'll keep it in [test-addusbtest](https://github.com/EmmanuelMess/AmazeFileManager/tree/test-addusbtest) though.

**Comment by TranceLove** [comment]
Never mind, I too have problem to plug in an external storage in my Robolectric tests when working on #1212.

Then you may have to write us a test case to prove your work is correct, since the problem description in #1175 is a bit vague.

**Comment by EmmanuelMess** [comment]
> Then you may have to write us a test case to prove your work is correct, since the problem description in #1175 is a bit vague.

But I can't:

> Okay, I removed the test, no idea how to make it work,

To replicate #1175:
1. connect USB OTG
2. select the device from the Drawer
3. set it up (select correct storage)
4. go to home
5. reopen the app via recents
6. select the device from the drawer

Notice how you have to set it up again. This PR fixes that.

**Comment by EmmanuelMess** [comment]
Added two tests, and a whole lot of reflection. Tests are not running on APIs 19 to 22 though, it seems the devs changed the constructor, will continue reflecting...

**Comment by EmmanuelMess** [comment]
Tests are now only not running on API 19, ideas?:

```
java.lang.RuntimeException: Could not find any resource from reference ResName{com.amaze.filemanager:style/Platform_V11_AppCompat_Light} from Style com.amaze.filemanager:Platform_V14_AppCompat_Light with theme with no applied styles

	at org.robolectric.res.StyleResolver.getParent(StyleResolver.java:99)
	at org.robolectric.res.StyleResolver.getAttrValue(StyleResolver.java:33)
	at org.robolectric.res.ThemeStyleSet.getAttrValue(ThemeStyleSet.java:17)
	at org.robolectric.shadows.ShadowAssetManager.findAttributeValue(ShadowAssetManager.java:996)
	at org.robolectric.shadows.ShadowAssetManager.buildTypedValue(ShadowAssetManager.java:882)
	at org.robolectric.shadows.ShadowAssetManager.attrsToTypedArray(ShadowAssetManager.java:930)
	at org.robolectric.shadows.ShadowResources$ShadowTheme.obtainStyledAttributes(ShadowResources.java:246)
	at android.content.res.Resources$Theme.obtainStyledAttributes(Resources.java)
	at android.content.Context.obtainStyledAttributes(Context.java:413)
	at android.support.v7.widget.TintTypedArray.obtainStyledAttributes(TintTypedArray.java:53)
	at android.support.v7.app.AppCompatDelegateImplBase.__constructor__(AppCompatDelegateImplBase.java:126)
	at android.support.v7.app.AppCompatDelegateImplBase.<init>(AppCompatDelegateImplBase.java)
	at android.support.v7.app.AppCompatDelegateImplV9.<init>(AppCompatDelegateImplV9.java)
	at android.support.v7.app.AppCompatDelegateImplV14.<init>(AppCompatDelegateImplV14.java)
	at android.support.v7.app.AppCompatDelegate.create(AppCompatDelegate.java:202)
	at android.support.v7.app.AppCompatDelegate.create(AppCompatDelegate.java:183)
	at android.support.v7.app.AppCompatActivity.getDelegate(AppCompatActivity.java:519)
	at android.support.v7.app.AppCompatActivity.onCreate(AppCompatActivity.java:70)
	at com.amaze.filemanager.activities.superclasses.PreferenceActivity.onCreate(PreferenceActivity.java:40)
	at com.amaze.filemanager.activities.superclasses.ThemedActivity.onCreate(ThemedActivity.java:28)
	at com.amaze.filemanager.activities.MainActivity.onCreate(MainActivity.java:275)
	at android.app.Activity.performCreate(Activity.java:5243)
	at org.robolectric.util.ReflectionHelpers.callInstanceMethod(ReflectionHelpers.java:249)
	at org.robolectric.android.controller.ActivityController$1.run(ActivityController.java:58)
	at org.robolectric.shadows.ShadowLooper.runPaused(ShadowLooper.java:365)
	at org.robolectric.android.controller.ActivityController.create(ActivityController.java:55)
	at org.robolectric.android.controller.ActivityController.create(ActivityController.java:65)
	at com.amaze.filemanager.filesystem.usb.UsbOtgTest.usbConnectionTest(UsbOtgTest.java:34)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:568)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:253)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:130)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:42)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:84)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at com.intellij.rt.execution.application.AppMainV2.main(AppMainV2.java:131)
```

File: app/src/test/java/com/amaze/filemanager/filesystem/usb/ReflectionHelpers.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    static UsbDevice callUsbDeviceConstructor(@NonNull String name, int vendorId, int productId,
+                                              int usbClass, int subClass, int protocol,
+                                              @NonNull Parcelable[] interfaces)
+            throws ClassNotFoundException, NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
+
+        Class<UsbDevice> clazz = (Class<UsbDevice>) Class.forName("android.hardware.usb.UsbDevice");
+        Constructor<UsbDevice> constructor = clazz.getConstructor(String.class,
+                int.class, int.class, int.class, int.class, int.class, Parcelable[].class);
+
+        return constructor.newInstance(name, vendorId, productId, usbClass, subClass, protocol);
```
Comment (TranceLove) [inline app/src/test/java/com/amaze/filemanager/filesystem/usb/ReflectionHelpers.java]: `interfaces` was omitted in calling the constructor, causing the test to fail at API 19 (not the exception you got though)

**Comment by EmmanuelMess** [comment]
> Exception: java.lang.NullPointerException thrown from the UncaughtExceptionHandler in thread "AsyncTask #1"

Weren't this type of exception what you fixed with `runOnUiThread()` in another PR?

File: app/src/test/java/com/amaze/filemanager/filesystem/usb/ReflectionHelpers.java (hunk 2 above)
Comment (EmmanuelMess) [reply app/src/test/java/com/amaze/filemanager/filesystem/usb/ReflectionHelpers.java]: Fixed!

**Comment by EmmanuelMess** [comment]
Updated Robolectric to 3.8 to remove the `maxSdk = 25` restriction

**Comment by TranceLove** [comment]
> Updated Robolectric to 3.8 to remove the maxSdk = 25 restriction

Please also update `org.robolectric:shadows-multidex` and `org.robolectric:shadows-httpclient` to 3.8.

**Comment by EmmanuelMess** [comment]
@TranceLove fixed

**Comment by EmmanuelMess** [comment]
Fixed

**Comment by VishalNehra** [comment]
Are you using a static object to persist the OTG state across app restarts?

**Comment by EmmanuelMess** [comment]
>  OTG state across app restarts?

Nope, across pause-resumes.

**Comment by VishalNehra** [comment]
So it does not persist OTG connection after closing  starting amaze after some time?
Can't ````SharedPreferences```` do the trick?

**Comment by EmmanuelMess** [comment]
> So it does not persist OTG connection after closing starting amaze after some time?

No.

> Can't `SharedPreferences` do the trick?

Yes, but I assumed a user that has closed the app that long would probably have removed the USB OTG anyway.

**Comment by TranceLove** [comment]
> Can't SharedPreferences do the trick?

> Yes, but I assumed a user that has closed the app that long would probably have removed the USB OTG anyway.

While it makes sense, some users may use OTG from time to time, using the same OTG cable and USB storage. It may become a good idea if the app can remember the storage permission this long.

(Although, one may use security concern to argue against this)

**Comment by EmmanuelMess** [comment]
> It may become a good idea if the app can remember the storage permission this long.

I think you misunderstand the fact that each time the user removes and reattaches a device we have to re-ask for the URI to Android, makes no sense saving the URI after a device has been detached

**Comment by EmmanuelMess** [comment]
> Doesn't matter how many times Amaze was closed/opened, we shouldn't ask them for permission again and again.

We don't, unless the user has closed Amaze for enough time for the system kill it, in which case I assume it is probable the user has removed the device anyway.

**Comment by EmmanuelMess** [comment]
@VishalNehra any thoughts?

**Pull request merged**